service/src/main/java/org/cbioportal/service/impl/GeneServiceImpl.java

```diff
@@ -31,50 +31,138 @@
+        List<Gene> geneList = geneRepository.getAllGenes(projection, pageSize, pageNumber, sortBy, direction);
+
+        for (Gene gene : geneList) {
+            gene.setChromosome(getChromosome(gene.getCytoband()));
         }
 
-        return genes;
+        return geneList;
     }
 
-    private String getChromosome(Gene gene) {
+    @Override
+    public BaseMeta getMetaGenes() {
 
-        if (gene.getCytoband() == null) {
+        return geneRepository.getMetaGenes();
+    }
+
+    @Override
+    public Gene getGene(String geneId) throws GeneNotFoundException {
+
+        Gene gene;
+
+        if (isInteger(geneId)) {
+            gene = geneRepository.getGeneByEntrezGeneId(Integer.valueOf(geneId));
+        } else {
+            gene = geneRepository.getGeneByHugoGeneSymbol(geneId);
+        }
+
+        if (gene == null) {
+            throw new GeneNotFoundException(geneId);
+        }
+
+        gene.setChromosome(getChromosome(gene.getCytoband()));
+        return gene;
+    }
+
+    @Override
+    public List<String> getAliasesOfGene(String geneId) {
+
+        if (isInteger(geneId)) {
+            return geneRepository.getAliasesOfGeneByEntrezGeneId(Integer.valueOf(geneId));
+        } else {
+            return geneRepository.getAliasesOfGeneByHugoGeneSymbol(geneId);
+        }
+    }
+
+    @Override
+    public List<Gene> fetchGenes(List<String> geneIds, String projection) {
+
+        List<Integer> entrezGeneIds = new ArrayList<>();
+        List<String> hugoGeneSymbols = new ArrayList<>();
+
+        splitIdsByType(geneIds, entrezGeneIds, hugoGeneSymbols);
+
+        List<Gene> geneList = geneRepository.fetchGenesByEntrezGeneIds(entrezGeneIds, projection);
+        geneList.addAll(geneRepository.fetchGenesByHugoGeneSymbols(hugoGeneSymbols, projection));
+
+        for (Gene gene : geneList) {
+            gene.setChromosome(getChromosome(gene.getCytoband()));
+        }
+
+        return geneList;
+    }
+
+
+
+    @Override
+    public BaseMeta fetchMetaGenes(List<String> geneIds) {
+
+        List<Integer> entrezGeneIds = new ArrayList<>();
+        List<String> hugoGeneSymbols = new ArrayList<>();
+
+        splitIdsByType(geneIds, entrezGeneIds, hugoGeneSymbols);
+
+        BaseMeta baseMeta = new BaseMeta();
+        baseMeta.setTotalCount(geneRepository.fetchMetaGenesByEntrezGeneIds(entrezGeneIds).getTotalCount() +
+                geneRepository.fetchMetaGenesByHugoGeneSymbols(hugoGeneSymbols).getTotalCount());
+
+        return baseMeta;
+    }
+
+    private void splitIdsByType(List<String> geneIds, List<Integer> entrezGeneIds, List<String> hugoGeneSymbols) {
+
+        for (String geneId : geneIds) {
+            if (isInteger(geneId)) {
+                entrezGeneIds.add(Integer.valueOf(geneId));
+            } else {
+                hugoGeneSymbols.add(geneId);
+            }
+        }
+    }
+
+    private boolean isInteger(String geneId) {
+        return geneId.matches("^-?\\d+$");
```

Comment:
yeah Apache Commons detects too much. Couldn't find one for simple integers with minus sign.